File: algorithms/2_medium/M/03779_minimum_number_of_operations_to_have_distinct_elements/vector.cpp
class Solution {
    static void normalize_the_vec(vector<int>& nums) {
        const int min_elem{ *ranges::min_element(nums) };

        for (auto& num : nums) {
            num -= min_elem;
        }
    }
public:
    static int minOperations(vector<int> nums) {
        const size_t len{ nums.size() };
        normalize_the_vec(nums);
        const int max_elem{ *ranges::max_element(nums) };
        vector<bool> seen(static_cast<size_t>(max_elem + 1), false);

        for (size_t i{ len - 1 }; i < len; --i) {
            if (seen[static_cast<size_t>(nums[i])]) {
                return static_cast<int>((i / 3) + 1);
            }
            seen[static_cast<size_t>(nums[i])] = true;
        }

        return 0;
    }
};

File: algorithms/2_medium/M/03779_minimum_number_of_operations_to_have_distinct_elements/vector.rs
impl Solution {
    fn normalize_the_vec(nums: &mut Vec<i32>) {
        let min_elem = nums.iter().map(|x| *x).min().unwrap(); 

        for num in nums.iter_mut() {
            *num -= min_elem;
        }
    }

    pub fn min_operations(mut nums: Vec<i32>) -> i32 {
        let len = nums.len();
        Self::normalize_the_vec(&mut nums);
        let max_elem = nums.iter().map(|x| *x).max().unwrap();
        let mut seen = vec![false; max_elem as usize + 1];

        for i in (0..len).rev() {
            let cache = &mut seen[nums[i] as usize];
            if *cache {
                return ((i / 3) + 1) as i32;
            }
            *cache = true;
        }

        0
    }
}
